[Nullables] wrap collection fields with nullable-support attributes in GetProto<T> and use .google.protobuf types for nullable fields

DeagleGross opened the pull request:
According to [nullwrappers doc](https://github.com/protobuf-net/protobuf-net/blob/main/docs/nullwrappers.md), we need to implement proto definition generation for `[SupportNull]`, `[NullWrappedValue]`

NickCraver commented on src/protobuf-net.Core/protobuf-net.Core.csproj:
```diff
@@ -11,7 +11,11 @@
+
+  <ItemGroup Condition="'$(TargetFramework)' == 'net462'">
+	<PackageReference Include="System.ValueTuple" Version="4.4.0" />
+  </ItemGroup>
```
FWIW, I would advise against adding this. System.ValueTuple creates a lot of friction on full framework with bindings and redirects, so much that we avoid it completely whenever possible. Given there's not a lot of usage here, it'd make life less painful for users if this wasn't a dependency.

mgravell replied:
@DeagleGross Nick is 100% right about this; friends to let friends use ValueTuple when there's any non-zero chance that you'll be on netfx - either explicitly (in this case) or indirectly (via netstandard)

DeagleGross replied:
Hey, thanks for pointing that out - did not know that - will remove that!

mgravell replied:
There's a bunch of 5 (Nick covers them here: https://nickcraver.com/blog/2020/02/11/binding-redirects/) where the runtime loader is just broken; they can't be used reliably on netfx

If you're an application developer, they're fine - you can just tweak the binding redirects once, job done

But if you're a library developer; it causes pain for *every* consumer; some % will know how to fix it, some % will give up, and some % will log support tickets

DeagleGross replied:
thanks for explaining (great article btw), removed ValueTuple nuget package

mgravell commented on src/protobuf-net.Test/NullWrappedValueTests.cs:
```diff
@@ -22,22 +22,6 @@ public NullWrappedValueTests(ITestOutputHelper log)
             => _log?.WriteLine(message);
 
-        [Fact]
-        public void ExistingListsBehaviour()
```
why are we removing this? is this moved, or...?

DeagleGross replied:
This was moved to `CollectionsWithNullsTests_Legacy`, since I decided to create a dedicated directory for tests over collections with nulls.

If you are not fine with that - please, re-open a comment and suggest a better place :)

mgravell replied:
That's fine. I had meant to delete this from the pending review notes after I found it later on, but I failed

mgravell commented on src/protobuf-net.Test/Nullables/Abstractions/CollectionsWithNullsTestsBase.cs:
```diff
@@ -0,0 +1,66 @@
+
+        protected virtual void Setup() { }
+
+        protected T SerializeAndDeserialize<T>(T instance)
```
probably not needed; unless you're using `ms` to check the payload, this is just `Serializer.DeepClone(instance)`, with extra steps

DeagleGross replied:
removed Serialize&Deserialize and using DeepClone now

mgravell commented on src/protobuf-net.Test/Nullables/CollectionsWithNullsTests_Legacy.cs:
```diff
@@ -0,0 +1,34 @@
+            using var ms = new MemoryStream(new byte[] { 0x08, 0x00, 0x08, 0x01, 0x08, 0x02 });
+            var clone = Serializer.Deserialize<Foo>(ms);
+            if (!ms.TryGetBuffer(out var buffer)) buffer = new ArraySegment<byte>(ms.ToArray());
+            Assert.Equal("08-00-08-01-08-02", BitConverter.ToString(buffer.Array, buffer.Offset, buffer.Count));
```
we didn't serialize, so what else do we expect? note: if this was in the code you copied from me: blame me!

DeagleGross replied:
removed legacy test completely - we have `CollectionsWithNullsTests_Basic`, which checks schema \ payload \ serialization behavior

mgravell commented on src/protobuf-net.Test/Nullables/CollectionsWithNullsTests_SupportNull.cs:
```diff
@@ -0,0 +1,79 @@
+
+        protected override void Setup()
+        {
+            RuntimeTypeModel.Default[typeof(SupportsNullListModel)][1].SupportNull = true;
```
note: if we want to test with+without, I would suggest using a bespoke model each time:

```
var model = RuntimeTypeModel.Create();
```

Note that you can't use `Serializer.Whatever(...)`, since `Serialize.*` is a proxy to `RuntimeTypeModel.Default.*`, but: `model.*` should work fine

This avoids tests interfering with eachother - concurrency, order, etc

DeagleGross replied:
Not to intercept with other test classes, using separate instance for each test class

mgravell commented on src/protobuf-net/Meta/MetaType.cs:
```diff
@@ -2222,8 +2244,40 @@ public bool IsGroup
+                    void WriteWrappedFieldPayload()
+                    {
+                        builder
+                            .Insert("group ", ref pos, condition: () => member.SupportNull || member.NullWrappedValueGroup)
```
I *think* (please check) that `SupportNull` already sets `NullWrappedValueGroup`, so you might not need to test both

DeagleGross replied:
Unfortunately, we have to check both
![image](https://user-images.githubusercontent.com/31598696/211208170-cf28f9f9-a50d-4957-abfc-30c7fd7d2907.png)

mgravell commented on src/protobuf-net/Meta/MetaType.cs:
```diff
@@ -2264,6 +2318,8 @@ void AppendReservations()
             }
         }
 
+        private static string GetWrappedSchemaTypeName(string schemaTypeName) => $"Wrapped{schemaTypeName}";
```
`"Wrapped" + schemaTypeName` please - more direct

DeagleGross replied:
just to learn: why is it better? I thought string concatenation is not a best option for performance - or do we rely on generated IL so much, that we will see something else there?

P.S. changed to yours proposition

NickCraver replied:
Ultimately, this is getting compiled (optimized) to the same thing ([example here](https://sharplab.io/#v2:D4AQzABCBMEMIQN4FgBQEMTZqkQEYAGCAMQHsyBjAQwBcAKA4gZ0oAsBTAW2oBUBPAA4cActS4cAlBAC8APmw5MAIgDqAJ2qDhAE2UQA1BFaceA4WIkBuRRlu4oRUhXU8GTY+258ho8VNk5LHQlDAASNU1tDh1EE29zPwkAX2UbELtUZKA==)) - so expressing it as a concat is more direct/accurate to what's happening :)

mgravell replied:
Darn it I was just sharplabing!  In this case it works out the same, so: fine. There are many cases where it doesn't. Now: that doesn't mean we shouldn't use it! There are many cases where it simplifies code greatly, but then the question is "how often does this run, do I care about extra overheads?". The main reason I said to prefer simple concat is just to reduce mental load, so I don't even need to ask myself that question.

mgravell reviewed (commented):
IIRC the tests only cover messages as contents, in terms of payload - I think we should also check what it generates (and serializes?) for `List<int>`, `List<int?>`, `List<string>`, etc 

Definitely looking good, 👍

mgravell commented:
Additional considerations:

- if writing proto2, we need "optional" in the wrapper message
- what happens if the same type is used multiple times in the model? Do we
write the wrapper message exactly once?

(Tests needed)

On Sat, 7 Jan 2023, 18:40 Nick Craver, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/protobuf-net.Core/protobuf-net.Core.csproj
> <https://github.com/protobuf-net/protobuf-net/pull/1001#discussion_r1064034216>
> :
>
> > @@ -11,7 +11,11 @@
>      <Description>Provides simple access to fast and efficient "Protocol Buffers" serialization from .NET applications</Description>
>      <!--<DefineConstants>$(DefineConstants);TRACK_USAGE</DefineConstants>-->
>    </PropertyGroup>
> -
> +
> +  <ItemGroup Condition="'$(TargetFramework)' == 'net462'">
> +	<PackageReference Include="System.ValueTuple" Version="4.4.0" />
> +  </ItemGroup>
>
> FWIW, I would advise against adding this. System.ValueTuple creates a lot
> of friction on full framework with bindings and redirects, so much that we
> avoid it completely whenever possible. Given there's not a lot of usage
> here, it'd make life less painful for users if this wasn't a dependency.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/protobuf-net/protobuf-net/pull/1001#pullrequestreview-1239696018>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEHMH4GVAHDURVT2OUQCDWRG2BRANCNFSM6AAAAAATUC2RSM>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

DeagleGross commented:
@mgravell if writing `proto2`, 
1) do we need `optional` keyword in any attribute scenario? 
2) do we need to include `group` as well?
```
message WrappedBar 
{
  optional Bar value = 1;
}
```

mgravell commented:
> @mgravell if writing `proto2`,
> 
> 1. do we need `optional` keyword in any attribute scenario?
> 2. do we need to include `group` as well?
> 
> ```
> message WrappedBar 
> {
>   optional Bar value = 1;
> }
> ```

Actually, I think it should always be `optional`, thinking about it. For whether it is group or not (in the as-group case): I'll need to check tomorrow (I need to look at the bytes)

DeagleGross commented:
TODO: 
- [x] remove legacy test
- [ ] place `optional` in every wrapped model group
- [ ] place `repeated group` inside of original model
- [ ] support differently typed fields

DeagleGross commented on src/protobuf-net.Test/NullListSupport.cs:
```diff
@@ -271,20 +258,6 @@ public void BasicPackedLists_PocoList(bool supportNull, Scenario scenario, Outco
-}");
-
-        [Fact]
-        public void BasicGroupedLists_Schema_SupportNull() => CheckSchema<BasicGroupedLists>(true, @"syntax = ""proto3"";
```
I have removed tests, because I have added similar ones in `Nullables` folder.
Sorry, did not see these tests existed here. Hope you are fine with that

DeagleGross commented:
@mgravell I thought about another possibly broken scenario by my changes. Consider this:
```
[ProtoContract]
class Model
{
    [ProtoMember(1), NullWrappedValue]
    public List<Test1.BarTest?> Items1 { get; set; } = new();

    [ProtoMember(2), NullWrappedValue]
    public List<Test2.BarTest?> Items2 { get; set; } = new();
}
```

Current PR's code generates such schema:
```
message BarTest {
   int32 Id = 1;
}
message BarTest {
   int32 Id = 1;
}
```
And schemas are not unique, due to namespace difference - so it is an invalid definition.

Question:
1) would it be fine to use `_valueMember.ItemType.FullName.Replace('.', '_')`  to calculate a schemaTypeName for `wrapped` model? Names will be long and non-readable, but it would be a "safe" approach of generation unique typeNames
2) if you want I can implement "semi-approach": if there would a collision of types found, then fully-qualified type names are used (`Test1.BarTest`); otherwise we can simply write schemaTypeName (`BarTest`)

Asking because Im not sure of what approach you would prefer! thanks!

mgravell commented:
looking good; re the name collision... yeah, that's a tricky one; numeric is a bad idea because the numbers could change arbitrarily and unpredictably (i.e. what was Foo1 can later become Foo5). Honestly, I think we should probably just leave it broken in this case - better for someone to fix it explicitly. I could get behind:

``` 
message BarTest {
   int32 Id = 1;
}
// warning: duplicate message name; you can use [ProtoContract(Name = "...")] to supply an alternative schema name
message BarTest {
   int32 Id = 1;
}
```

thoughts?

DeagleGross commented:
@mgravell, I like your idea about "warning for duplication of message names", **implemented it, please, check latest changes**!

To provide more details: if `schemaTypeName` is one of _well-known_ types (i.e. `int32` or `string` or ... - basically any commonly used System.XXX types) I dont submit such a warning, because IMO it does not make sense to write a warning, if your `ProtoContract` has 2 collections with `int` as properties:
```
[ProtoContract]
public class Contract
{
   [ProtoMember(1), NullWrappedValue]
   public List<int> X { get; set; }

   [ProtoMember(2), NullWrappedValue]
   public List<int> Y { get; set; }
}
```
schema:
```
@"syntax = ""proto3"";

// // do we need this comment? I dont think so
// warning: duplicate message name; you can use [ProtoContract(Name = ""..."")] to supply an alternative schema name
message Wrappedint32 {
   optional int32 value = 1;
}
message Contract {
   repeated Wrappedint32 X = 1;
   repeated Wrappedint32 Y = 2;
}
```

mgravell commented:
Well, if the type is the same, we only emit the message once anyway, don't we? so... no duplicate?

however, a really interesting question is what happens here:

``` c#
[ProtoContract]
public class Contract
{
   [ProtoMember(1), NullWrappedValue]
   public List<SomeOtherType> X { get; set; }

   [ProtoMember(2), NullWrappedValue(AsGroup = true)]
   public List<SomeOtherType> Y { get; set; }
}
```

these then *cannot* be the same message, since they have different layout; some disambiguation needed? Maybe as simple as `WrappedSomeOtherType` vs `WrappedAsGroupSomeOtherType` ?

DeagleGross commented:
@mgravell 
1. Re: `Well, if the type is the same, we only emit the message once anyway, don't we? so... no duplicate?` - as discussed, we dont want to generate different wrappers for `int` & `int?`, thats why I added a check on that.

2. Re: `disambiguation of differently null-wrapped properties` - Do you want me to generate different names **always** or **only in case of ambiguity**? Easier change would be to always submit unique type names for wrapped models (_something like this:_)
```
public string WrappedSchemaTypeName
{
    get
    {
        var typeName = !string.IsNullOrEmpty(_alternativeTypeName)
            ? _alternativeTypeName
            : _originalSchemaTypeName;

        if (ValueMember.SupportNull) return "WrappedAsSupportNull" + typeName;
        if (ValueMember.NullWrappedValueGroup) return "WrappedAsGroup" + typeName;
        if (ValueMember.NullWrappedValue) return "WrappedAsNWV" + typeName;
        return "Wrapped" + typeName;
    }
}
```

mgravell commented:
Always is better. So the name doesn't flip on an existing type just by
adding more types.

On Tue, 24 Jan 2023, 18:36 Korolev Dmitry, ***@***.***> wrote:

> @mgravell <https://github.com/mgravell>
>
>    1.
>
>    Re: Well, if the type is the same, we only emit the message once
>    anyway, don't we? so... no duplicate? - as discussed, we dont want to
>    generate different wrappers for int & int?, thats why I added a check
>    on that.
>    2.
>
>    Re: disambiguation of differently null-wrapped properties - Do you
>    want me to generate different names *always* or *only in case of
>    ambiguity*? Easier change would be to always submit unique type names
>    for wrapped models (*something like this:*)
>
> public string WrappedSchemaTypeName
> {
>     get
>     {
>         var typeName = !string.IsNullOrEmpty(_alternativeTypeName)
>             ? _alternativeTypeName
>             : _originalSchemaTypeName;
>
>         if (ValueMember.SupportNull) return "WrappedAsSupportNull" + typeName;
>         if (ValueMember.NullWrappedValueGroup) return "WrappedAsGroup" + typeName;
>         if (ValueMember.NullWrappedValue) return "WrappedAsNWV" + typeName;
>         return "Wrapped" + typeName;
>     }
> }
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/protobuf-net/protobuf-net/pull/1001#issuecomment-1402412534>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEHMDOTR5L2SQ7RGDIZKTWUAOJ5ANCNFSM6AAAAAATUC2RSM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

DeagleGross commented:
Sorry for delay - I've refactored code a bit to support all corner cases more easily.
PR is ready to be re-reviewed!

DeagleGross commented:
I have also included changes needed to support `wrappers.proto` parsing in last 2 commits. 
So, current PR contains:
- schema generation support for inner types of `List<T>` fields, which wraps type into a separate message model (for `SupportsNull` \ `NullWrappedValue(AsGroup = true)` \ `NullWrappedValue`
- schema generation for nullable well-known types (like `int?`), which are defined in `google.protobuf.wrappers.proto` namespace
- c# code generation support for nullable well-known types like `int?` which are defined in `google.protobuf.wrappers.proto` namespace
- and tests for all of the functionality

DeagleGross commented:
As discussed, have done these changes:
- added `group` modifier to field for wrappers.proto-compatible fields, which are marked with `NullWrappedValue(AsGroup)` and `SupportsNull`
- added `NullWrappedValue` (for wrappers proto) and `CompatibilityLevel.300` (for `DateTime` and `TimeSpan`) for CSharpCodeGeneration
- added flags `nullwrappers` and `compatibilityLevel` which allow to control behavior, if current output is desired to be kept

mgravell commented on src/protogen/Program.cs:
```diff
@@ -340,6 +340,10 @@ private static void ShowHelp()
   +oneof={default|enum}       Specify whether 'oneof' should generate enums.
   +listset={yes|no}           Specify whether lists should emit setters
   +services={grpc;wcf}        Semi-colon list of service metadata to support.
+  +nullwrappers={yes;no}      Specify whether wrappers.proto should be generated
```
yes|no for both of these

DeagleGross replied:
changed

mgravell commented on src/BuildToolsSmokeTests/Checks.cs:
```diff
@@ -0,0 +1,51 @@
+        [InlineData(typeof(Vanilla.Foo), nameof(Vanilla.Foo.I), typeof(double?))]
+        [InlineData(typeof(Enabled.Foo), nameof(Enabled.Foo.I), typeof(double?))]
+        [InlineData(typeof(Disabled.Foo), nameof(Disabled.Foo.I), typeof(DoubleValue))]
+        public void CheckPropertyType(Type source, string name, Type valueType)
```
can you add an additional test here, that checks (at runtime) whether the property is marked with the expected attribute? (and check you can get it working locally; you might need to change the Version in the csproj)

DeagleGross replied:
Added tests on all `wrappers.proto` types, but for some unknown to me reason build of project fails:
```
1>C:\Code\forked\protobuf-net\src\BuildToolsSmokeTests\Checks.cs(48,63,48,74): error CS0117: 'Foo' does not contain a definition for 'stringItems'      
```
here is a line where it fails
![image](https://user-images.githubusercontent.com/31598696/218263448-e1105827-6f2b-486f-b99c-07dee4acb62f.png)

and I see property exists
![image](https://user-images.githubusercontent.com/31598696/218263457-b7c9f2c3-4d01-4dc5-9670-9b57bcee4292.png)

Can you, please, take a look - what am I doing wrong?

mgravell replied:
Are you using the console and `dotnet test`? The IDE hates this setup

DeagleGross replied:
same thing. I did not do any change to analyzer, but rebuilt using your readme and still getting this error
![image](https://user-images.githubusercontent.com/31598696/218264348-dd19805d-abdd-4ab5-a585-7895f2264969.png)

DeagleGross replied:
silly me - for some reason .proto field was not changed.
It is fixed and works!

mgravell commented on src/protobuf-net.Reflection/CSharpCodeGenerator.cs:
```diff
@@ -839,23 +868,76 @@ protected string MakeRelativeName(GeneratorContext ctx, string typeName)
+                {
+                    case WellKnownTypeDouble:
+                        nullabilityType = NullabilityType.NullWrappedValue;
+                        return nonNullable ? "double" : "double?";
```
in what scenario would be emit `double` (not `double?`) for a `DoubleValue` ? do we have tests for that scenario?

DeagleGross replied:
that can never happen, agree. Returning `double?` no matter if the code flow passes `nonNullable` as `true` of `false`

mgravell commented on src/protobuf-net.Reflection/CSharpCodeGenerator.cs:
```diff
@@ -839,23 +868,76 @@ protected string MakeRelativeName(GeneratorContext ctx, string typeName)
+                
+                #endregion
+
+                #region BCL
```
I'm not sure we're adding much with a `#region` here; can we just `// BCL` and `// well-known types` ?

DeagleGross replied:
leftover after previous implementation approach, removed

mgravell reviewed (commented):
some nits and questions, but generally: looking great, thanks

mgravell commented:
see test failure above; look at  Build.csproj, the line `<ProjectReference Include="tests\**\*.csproj" />` - you should be able to add an `Exclude` on that new project (for now)

DeagleGross commented on Build.csproj:
```diff
@@ -14,8 +14,8 @@
+    <ProjectReference Include="src\*\*.csproj" Exclude="src\protogen.Site\*.csproj;src\BuildToolsSmokeTests\*.csproj" />
     <!-- <ProjectReference Include="samples\**\*.csproj" Exclude="samples\**\*Mvc5*.csproj" /> -->
-    <ProjectReference Include="tests\**\*.csproj" Exclude="src\BuildToolsSmokeTests\*.csproj" />
+    <!-- <ProjectReference Include="tests\**\*.csproj" /> -->
```
You did not screw the path (I did not check the folder structure before except looking at sln) -> there is no `tests\` directory at all.
So I have added exclude to `src` line, will see how it goes. Sorry for multiple commits - I dont know a better way to test a CI than "send a change and see what happens" :D

mgravell replied:
No problem whatsoever re commits; for CI that's the only strategy I know, too. And we can squash at merge. Let's see how the CI goes this time.

DeagleGross replied:
so build finally succeeded (some timeouts in CI, but I guess that happens), so we are good now

mgravell commented:
minor addition; can you look at the blazor changes (2 files) here? https://github.com/protobuf-net/protobuf-net/pull/856/files

mgravell commented:
CI fail looks to be an intermittent timeout we see in PEVerify; I'm re-running

DeagleGross commented:
@mgravell, re `minor addition; can you look at the blazor changes (2 files) here? https://github.com/protobuf-net/protobuf-net/pull/856/files`:
let me clarify your intention: you want me to add code, which sets `nullwrappers` & `compatlevel` flags (as a checkboxes in UI), yep? I have set checkboxes to "Disable xxx", because we are changing the default behavior.

UI:
![image](https://user-images.githubusercontent.com/31598696/218282172-a42581ec-58ba-41b0-a109-c6e46180570d.png)

Output with disabled turned on (current behavior):
![image](https://user-images.githubusercontent.com/31598696/218282184-0c0bdc63-77f6-4d87-ac09-7d9e68e8402f.png)

Output without marking the checkboxes (new default behavior):
![image](https://user-images.githubusercontent.com/31598696/218282198-67d09479-3206-4e8e-83b5-a2a5c2793fa5.png)

mgravell commented:
That's great, thanks. Looking good 👍. Happy with the state? Anything left to do?

DeagleGross commented:
I dont think so, seems like a finished PR

mgravell commented:
also 62 warnings in the null test?

DeagleGross commented:
Oh, sorry, did not see that. Will take a look tomorrow

DeagleGross commented:
4 warnings left, I guess they are not related to my changes. Sorry for inattention

mgravell commented:
Awesome, thanks (I believe I already have a PR that fixes the others; I didn't want to merge it ahead of you and put the merge on your shoulders)

mgravell commented:
Great bit of work, much appreciated!
